lager/lenses.hpp

```diff
@@ -123,51 +128,116 @@ auto attr(Member member)
+                    try {
+                        r.at(key) = std::move(part).value();
+                    } catch (std::out_of_range const&) {}
                 }
-            }())([&](auto&& x) {
-                auto r = std::forward<decltype(p)>(p);
-                try {
-                    r.at(key) = std::forward<decltype(x)>(x);
-                } catch (std::out_of_range const&) {}
                 return r;
             });
         };
     });
 }
 
+/*!
+ * `Key -> Lens<{X}, [X]>`
+ */
 template <typename Key>
-auto at_i(Key key)
-{
-    return zug::comp([=](auto&& f) {
-        return [f, &key](auto&& p) {
-            return f([&] {
+auto at_i(Key key) {
+    return zug::comp([key](auto&& f) {
+        return [f, &key](auto&& whole) {
+            using Part = std::optional<std::decay_t<decltype(whole.at(key))>>;
+
+            return f([&]() -> Part {
                 try {
-                    return std::forward<decltype(p)>(p).at(key);
-                } catch (std::out_of_range const&) {
-                    return std::decay_t<decltype(p.at(key))>{};
-                }
-            }())([&](auto&& x) {
-                if (static_cast<std::size_t>(key) < p.size()) {
-                    return std::forward<decltype(p)>(p).set(
-                        key, std::forward<decltype(x)>(x));
+                    return std::forward<decltype(whole)>(whole).at(key);
+                } catch (std::out_of_range const&) { return std::nullopt; }
+            }())([&](Part part) {
+                if (part.has_value() &&
+                    static_cast<std::size_t>(key) < whole.size()) {
+                    return std::forward<decltype(whole)>(whole).set(
+                        key, std::move(part).value());
                 } else {
-                    return std::forward<decltype(p)>(p);
+                    return std::forward<decltype(whole)>(whole);
                 }
             });
         };
     });
 }
 
+/*!
+ * `X -> Lens<[X], X>`
+ */
+template <typename T>
+auto fallback(T&& t) {
+    return zug::comp([t = std::forward<T>(t)](auto&& f) {
+        return [&, f](auto&& whole) {
+            return f(LAGER_FWD(whole).value_or(std::move(t)))(
+                [&](auto&& x) { return LAGER_FWD(x); });
+        };
+    });
+}
+
+/*!
+ * `() -> Lens<[X], X>`
+ */
+auto fallback() {
+    return zug::comp([](auto&& f) {
+        return [&, f](auto&& whole) {
+            using T = std::decay_t<decltype(whole.value())>;
+            return f(LAGER_FWD(whole).value_or(T{}))(
+                [&](auto&& x) { return LAGER_FWD(x); });
+        };
+    });
+}
+
+/*!
+ * `(Lens<W, P> | Lens<W, [P]>) -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optlift(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f](auto&& whole) {
+            using Whole = std::decay_t<decltype(whole)>;
+            using Part  = std::optional<std::decay_t<decltype(::lager::view(
```

Comment:
Oops, this actually only does `Lens<W, P> -> Lens <[W], [P]>`.
Should I make two separate functions for `Lens<W, [P]> -> Lens <[W], [P]>` and `Lens<W, P> -> Lens <[W], [P]>` or just keep them merged as one function.
Or even provide `optlift` as a convenience for both, and add two others for explicitly handling one or the other?

Reply:
Went with all three, although I'm weary of `optlift` being too much of a footgun.